expr/limit.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2018 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expr
+
+import (
+	"encoding/binary"
+	"errors"
+
+	"github.com/google/nftables/binaryutil"
+	"github.com/mdlayher/netlink"
+	"golang.org/x/sys/unix"
+)
+
+// Theses errors represent the possible parsing error of the rate limit.
+var (
+	ErrLimitTimeUnitInvalid = errors.New("expr: invalid limit time unit")
+	ErrLimitTypeInvalid     = errors.New("expr: invalid limit type")
+	ErrLimitUnhandledType   = errors.New("expr: unhandled limit type")
+)
+
+// LimitType represents the type of the limit expression.
+type LimitType uint32
+
+// Imported from the nft_limit_type enum in netfilter/nf_tables.h.
+const (
+	LimitTypePkts     LimitType = unix.NFT_LIMIT_PKTS
+	LimitTypePktBytes LimitType = unix.NFT_LIMIT_PKT_BYTES
+)
+
+func limitType(t uint32) (LimitType, error) {
+	switch LimitType(t) {
+	case LimitTypePkts:
+		return LimitTypePkts, nil
+	case LimitTypePktBytes:
+		return LimitTypePktBytes, nil
+	default:
+		return 0, ErrLimitTypeInvalid
+	}
+}
+
+// LimitTime represents the limit in number of seconds
+type LimitTime uint32
+
+// Possible limit unit values
+const (
+	LimitTimeSecond LimitTime = 1
+	LimitTimeMinute LimitTime = 60
+	LimitTimeHour   LimitTime = 60 * 60
+	LimitTimeDay    LimitTime = 60 * 60 * 24
+	LimitTimeWeek   LimitTime = 60 * 60 * 24 * 7
+)
+
+func limitTime(value uint64) (LimitTime, error) {
+	switch LimitTime(value) {
+	case LimitTimeSecond:
+		return LimitTimeSecond, nil
+	case LimitTimeMinute:
+		return LimitTimeMinute, nil
+	case LimitTimeHour:
+		return LimitTimeHour, nil
+	case LimitTimeDay:
+		return LimitTimeDay, nil
+	case LimitTimeWeek:
+		return LimitTimeWeek, nil
+	default:
+		return 0, errors.New("nftables: invalid limit value")
+	}
+}
+
+// Limit represents a rate limit expression
+type Limit struct {
+	Type  LimitType
+	Rate  uint64
+	Over  bool
+	Unit  LimitTime
```

Comment:
Why not make Unit a time.Duration and convert it in marshal()? That would feel more familiar to users, I think.